presto-main/src/main/java/com/facebook/presto/operator/aggregation/estimatendv/NDVEstimatorFunction.java

```diff
@@ -13,42 +13,129 @@
+                inputParameterMetadata,
+                inputFunction,
+                COMBINE_FUNCTION,
+                OUTPUT_FUNCTION,
+                ImmutableList.of(new AggregationMetadata.AccumulatorStateDescriptor(
+                        NDVEstimatorState.class,
+                        stateSerializer,
+                        new HistogramStateFactory(type, EXPECTED_SIZE_FOR_HASHING, HISTOGRAM_GROUP_MODE, true))),
+                BigintType.BIGINT);
+
+        Class<? extends Accumulator> accumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
+                Accumulator.class,
+                metadata,
+                classLoader);
+        Class<? extends GroupedAccumulator> groupedAccumulatorClass = AccumulatorCompiler.generateAccumulatorClass(
+                GroupedAccumulator.class,
+                metadata,
+                classLoader);
+        return new BuiltInAggregationFunctionImplementation(NAME, inputTypes, ImmutableList.of(intermediateType),
+                BigintType.BIGINT, true, false, metadata, accumulatorClass, groupedAccumulatorClass);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE),
+                new ParameterMetadata(BLOCK_INPUT_CHANNEL, type),
+                new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(Type type, NDVEstimatorState state, Block key, int position)
     {
-        state.add(freq);
+        Histogram.input(type, state, key, position);
     }
 
-    @CombineFunction
-    public static void combine(
-            @AggregationState NDVEstimatorState state,
-            @AggregationState NDVEstimatorState otherState)
+    public static void combine(NDVEstimatorState state, NDVEstimatorState otherState)
     {
-        state.merge(otherState);
+        Histogram.combine(state, otherState);
     }
 
-    @OutputFunction("bigint")
-    public static void output(@AggregationState NDVEstimatorState state, BlockBuilder out)
+    public static void output(NDVEstimatorState state, BlockBuilder out)
     {
         BigintType.BIGINT.writeLong(out, state.estimate());
     }
+
+    @Override
+    public BuiltInAggregationFunctionImplementation specialize(BoundVariables boundVariables, int arity, FunctionAndTypeManager functionAndTypeManager)
+    {
+        Type keyType = boundVariables.getTypeVariable("K");
+        Type intermediateMapType = functionAndTypeManager.getParameterizedType(StandardTypes.MAP, ImmutableList.of(
+                TypeSignatureParameter.of(keyType.getTypeSignature()),
+                TypeSignatureParameter.of(BigintType.BIGINT.getTypeSignature())));
+        return generateAggregation(keyType, intermediateMapType);
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Implement Chao estimator to estimate ndv based on sample";
```

Comment:
```suggestion
        return "an estimate of distinct values in a larger dataset using a subset of the data";
```

The part about using the Chao estimator can be written into the javadoc of this class.